app/coffee/WebsocketServer.js

```diff
@@ -0,0 +1,435 @@
+const { bootstrapSecret, pollingAuthSecret } = Settings.security
+
+const MISSING_BOOTSTRAP = new Error('missing bootstrap')
+const RESTART_BOOTSTRAP = new Error('restart bootstrap')
+const SESSION_LOOKUP_FAILED = new Error('could not look up session by key')
+function getParams(req, callback) {
+  const raw = req.url.slice(req.url.indexOf('bootstrap=') + 10)
+  if (!raw) {
+    metrics.inc('wss.bootstrap.failed', 1, { status: 'missing' })
+    return callback(MISSING_BOOTSTRAP)
+  }
+  const bootstrap = signature.unsign(raw, bootstrapSecret)
+  if (bootstrap === false) {
+    metrics.inc('wss.bootstrap.failed', 1, { status: 'invalid' })
+    return callback(RESTART_BOOTSTRAP)
+  }
+  const args = bootstrap.split(':')
+  const version = args.shift()
+  if (version !== 'v1') {
+    metrics.inc('wss.bootstrap.failed', 1, { status: 'old-v' })
+    return callback(RESTART_BOOTSTRAP)
+  }
+  const token = args.shift()
+  sessionRedisClient.get(`token:${token}`, function(error, sid) {
+    if (error) {
+      metrics.inc('wss.bootstrap.failed', 1, { status: 'redis' })
+      return callback(error)
+    }
+    if (!sid) {
+      metrics.inc('wss.bootstrap.failed', 1, { status: 'expired' })
+      return callback(RESTART_BOOTSTRAP)
+    }
+    sessionRedisClient.get(`sess:${sid}`, function(error, raw) {
+      if (error) {
+        metrics.inc('wss.bootstrap.failed', 1, { status: 'redis' })
+        return callback(error)
+      }
+      if (!raw) {
+        metrics.inc('wss.bootstrap.failed', 1, { status: 'logged-out' })
+        return callback(SESSION_LOOKUP_FAILED)
+      }
+      metrics.inc('wss.bootstrap.success', 1)
+      callback(null, JSON.parse(raw), ...args)
+    })
+  })
+}
+function sessionStart(req, client, callback) {
+  getParams(req, function(error, ...args) {
+    client.on('disconnect', () => clientMap.delete(client.id))
+    clientMap.set(client.id, client)
+    clientEmitter.emit('client', error, client, ...args)
+    callback()
+  })
+}
+
+const ACCESS_CONTROL_ALLOWED_ORIGINS = Settings.socketIoOrigins.split(',')
+function addCORSHeaders(req, res, next) {
+  if (ACCESS_CONTROL_ALLOWED_ORIGINS.includes(req.headers.origin)) {
+    res.setHeader('Access-Control-Allow-Origin', req.headers.origin)
+  }
+  res.vary('Origin')
+  next()
+}
+
+const clientMap = new Map()
+const clientEmitter = new EventEmitter()
+module.exports.clientMap = clientMap
+module.exports.clientEmitter = clientEmitter
+module.exports.attachRoutes = function(app) {
+  app.get(`${Settings.socketIoPath}/socket.io.js`, function(req, res) {
+    res.contentType('application/javascript')
+    res.send('window.io="plain"')
+  })
+
+  const POLLING_PREFIX = `${Settings.socketIoPath}/deprecated-polling`
+  app.use(`${POLLING_PREFIX}/`, addCORSHeaders)
+  app.post(`${POLLING_PREFIX}/new`, function createNewSession(req, res) {
+    const polling = new PollingWebSocketShim()
+    const client = new Client(polling, clientMap)
```

Comment:
This constructor has two arguments, but the `Client` class in this file only appears to take one. Is that right?

Reply:
Yes, the second parameter is an artifact from a previous iteration and can get removed.

_The `Client` wrapper was tasked to set up the client tracking (insert into the `clientMap` and register a `disconnect` handler to remove it). This logic resides [in the session start](https://github.com/overleaf/real-time/blob/b83405de377b448f9d1a64d17a78552109ee8e96/app/coffee/WebsocketServer.js#L59-L60) now._